src/components/withCurrentUserPersonalDetails.tsx

```diff
@@ -0,0 +1,63 @@
+import React, {ComponentType, RefAttributes, ForwardedRef, useMemo} from 'react';
+import {OnyxEntry, withOnyx} from 'react-native-onyx';
+import getComponentDisplayName from '../libs/getComponentDisplayName';
+import ONYXKEYS from '../ONYXKEYS';
+import personalDetailsPropType from '../pages/personalDetailsPropType';
+import type {PersonalDetails, Session} from '../types/onyx';
+
+type CurrentUserPersonalDetails = PersonalDetails | Record<string, never>;
+
+type WithCurrentUserPersonalDetailsProps = {
+    currentUserPersonalDetails: CurrentUserPersonalDetails;
+};
+
+type WithCurrentUserPersonalDetailsComponentProps = {
+    /** Personal details of all the users, including current user */
+    personalDetails: Record<string, PersonalDetails> | null;
+
+    /** Session of the current user */
+    session: OnyxEntry<Session>;
+};
+
+// TODO: remove when all components that use it will be migrated to TS
+const withCurrentUserPersonalDetailsPropTypes = {
+    currentUserPersonalDetails: personalDetailsPropType,
+};
+
+const withCurrentUserPersonalDetailsDefaultProps: WithCurrentUserPersonalDetailsProps = {
+    currentUserPersonalDetails: {},
+};
+
+export default function <TProps extends WithCurrentUserPersonalDetailsProps, TRef>(WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>) {
+    function WithCurrentUserPersonalDetails(props: Omit<TProps, keyof WithCurrentUserPersonalDetailsProps> & WithCurrentUserPersonalDetailsComponentProps, ref: ForwardedRef<TRef>) {
+        const accountID = props.session?.accountID ?? 0;
+        const accountPersonalDetails = props.personalDetails?.[accountID];
+        const currentUserPersonalDetails: CurrentUserPersonalDetails = useMemo(
+            () => (accountPersonalDetails ? {...accountPersonalDetails, accountID} : {}),
+            [accountPersonalDetails, accountID],
+        );
+        return (
+            <WrappedComponent
+                // eslint-disable-next-line react/jsx-props-no-spreading
+                {...(props as TProps & WithCurrentUserPersonalDetailsComponentProps)}
+                ref={ref}
+                currentUserPersonalDetails={currentUserPersonalDetails}
+            />
+        );
+    }
+
+    WithCurrentUserPersonalDetails.displayName = `WithCurrentUserPersonalDetails(${getComponentDisplayName(WrappedComponent)})`;
+
+    const withCurrentUserPersonalDetails = React.forwardRef(WithCurrentUserPersonalDetails);
+
+    return withOnyx<Omit<TProps, keyof WithCurrentUserPersonalDetailsProps> & WithCurrentUserPersonalDetailsComponentProps, WithCurrentUserPersonalDetailsComponentProps>({
```

Comment:
Looking back at `withOnyx` type declaration I think we mistyped it.

Return type isn't generic enough and we lose the ref type because of it:
```ts
declare function withOnyx<TComponentProps, TOnyxProps>(
    mapping: {
        [TOnyxProp in keyof TOnyxProps]:
            | OnyxPropMapping<TComponentProps, TOnyxProps, TOnyxProp>
            | OnyxPropCollectionMapping<TComponentProps, TOnyxProps, TOnyxProp>;
    },
    shouldDelayUpdates?: boolean,
): (component: React.ComponentType<TComponentProps>) => React.ComponentType<Omit<TComponentProps, keyof TOnyxProps>>;
```

Return type should be something like this:
```ts
declare function withOnyx<TComponentProps, TOnyxProps>(
    mapping: {
        [TOnyxProp in keyof TOnyxProps]:
            | OnyxPropMapping<TComponentProps, TOnyxProps, TOnyxProp>
            | OnyxPropCollectionMapping<TComponentProps, TOnyxProps, TOnyxProp>;
    },
    shouldDelayUpdates?: boolean,
): <TRef>(component: (props: TComponentProps & React.RefAttributes<TRef>) => React.ReactElement | null) => (props: Omit<TComponentProps, keyof TOnyxProps> & React.RefAttributes<TRef>) => React.ReactElement | null

```

@VickyStash @kubabutkiewicz wdyt?